File: Tarea1b/SRC/Haskell/ejercicios.hs
--MODELADO Y PROGRAMACIÓN 2019-2
--Tarea 1b
--Author: Martin Felipe Espinal Cruces.
--Date 2019-02-14

{-
Función que dado dos números enteros y la ayuda de dos funciones
auxiliares "minimo comun" y "multiplos", regresa un entero que 
representa el mínimo común múltiplo.
-}
mcm :: Int -> Int -> Int
mcm 0 _ = 0
mcm _ 0 = 0
mcm a b
   | a == b = a
   | otherwise = minimoComun (multiplos a) (multiplos b)

{-
Función auxiliar que recibe como parámetro dos listas de 
entero verificando valor por valor el elemento mas pequeño
de ambas listas, y lo devuleve.
-}
minimoComun :: [Int] -> [Int] -> Int
minimoComun (x:xs) (y:ys)
   | x > y = minimoComun (x:xs) ys
   | x < y = minimoComun (y:ys) xs
   | otherwise = x


{-
Función auxiliar que dado un número entero, devuelve una
lista de enteros que son divisores comunes de un número.
-}
multiplos :: Int -> [Int]
multiplos n = [x*n | x <- [1,2..]]


{-
Función que recibe dos valores de tipo entero que ayudandose
de dos funciones auxiliares "divisoresPropios" y "comun", y que devuelve 
un entero con el máximo común divisor.
-}
mcd :: Int -> Int -> Int
mcd 0 _ = 0
mcd _ 0 = 0
mcd a b
   |a == b = a
   |otherwise = maximum $ comun (divisoresPropios a) (divisoresPropios b)

{-
Función auxiliar que recibe dos listas y regresa una lista con los
elementos en común.
-}
comun :: Eq a => [a] -> [a] -> [a]
comun a b = filter (\x -> x `elem` a) b

{-
Función auxiliar que dado dos enteros regresa una lista de enteros en la
que se encuentran los valores que dividen al entero.
-}
divisores :: Int -> Int -> [Int]
divisores n 1 = [1]
divisores n m
   | (mod n m == 0) = (m):divisores n (m-1) 
   | otherwise = divisores n (m-1)

{-
Lista auxiliar por comprensión que devuelve la lista de todos divisores de un número
-}
divisoresPropios n = [x | x <-divisores n n, x/=n || x == n]

{-
Función que verifica si el último elemento de la sucesión de Collatz es 1 y devuelve 
verdadero en caso de ser así, o falso en caso contrario.
-}
collatz :: Int -> Bool
collatz 0 = False
collatz 1 = True
collatz a 
   | (last (sucesion a) == 1) = True
   | otherwise = False

{-
Función auxiliar a collatz que regresa una lista con la sucesión de Collatz de un número.
-}
sucesion :: Int -> [Int]
sucesion 1 = []
sucesion y
   | (y `mod` 2 == 0) = par : sucesion par
   | otherwise = impar : sucesion impar
   where 
    par = y `div` 2
    impar = (y*3) +1

{-
Función auxiliar que dado un entero, genera una lista por comprención la cual recibe un una
lista de 2 hasta el número enviado (exceptando al 1 que no es par) y con la condición de saber
si el número de la lista es primo, respuesta que se obtiene con la función auxiliar "primos".
-}
factores ::Int -> [Int]
factores n  = [x | x <- [2..n] , primos x]

{-
Función auxiliar que dado un entero verifica si es un número primo.
-}
primos ::Int -> Bool
primos n = length (divisoresPropios n) <=2

File: Tarea1b/SRC/c++/ejercicios.cpp
/**
* @file ejercicios.cpp
* @Author Martin Felipe Espinal Cruces
* @date 09/02/2019
*/
#include <iostream>
#include <vector>


using namespace std;

int mcm(int num1, int num2);
int cantidad;
int mcd(int num1, int num2);
bool sucesionCollatz(int n);
int * factoresPrimos(int n);
void menu();

/**
* @brief Método principal donde se realizaran las pruebas.
* @param void
* @return int 0;
*/
int main() {
	bool continua = true;
	int opcion, numero1, numero2, seguir, factores, resultado;
	vector<int> respuesta;
	menu();
	cin >> opcion;
	while (continua){
		switch (opcion) {
		case 1: 
			cout << "introduce el primer número:" << endl;
			cin >> numero1;
			cout << "introduce el segundo número:" << endl;
			cin >> numero2;
			resultado = mcm(numero1, numero2);
			cout << resultado << endl;
			cout << "¿Desea regresar al menú principal?" << "Si (1), Salir (2)" << endl;
			cin >> seguir;
			if (seguir == 2) {
				continua  = false;
			}else if(seguir == 1) {
				menu();
				cin >> opcion;
			}
		break;

		case 2:
			cout << "introduce el primer número:" << endl;
			cin >> numero1;
			cout << "introduce el segundo número:" << endl;
			cin >> numero2;
			resultado = mcm(numero1, numero2);
			cout << resultado << endl;
			cout << "¿Desea regresar al menú principal?" << "Si (1), Salir (2)" << endl;
			cin >> seguir;
			if (seguir == 2) {
				continua  = false;
			}else if(seguir == 1) {
				menu();
				cin >> opcion;
			}
		break;

		case 3:
			cout << "introduce un número:" << endl;
			cin >> numero1;
			if (sucesionCollatz(numero1)) {
				cout << "El último numero de la sucesion es uno" << endl;
			}else {
				cout << "El último numero de la sucesion no es uno" << endl;
			}
			
			cout << "¿Desea regresar al menú principal?" << "Si (1), Salir (2)" << endl;
			cin >> seguir;
			if (seguir == 2) {
				continua  = false;
			}else if(seguir == 1) {
				menu();
				cin >> opcion;
			}
		break;

		case 4:
			cout << "Elige un número" << endl;
			cin >> factores;
			static int * r = factoresPrimos(factores);
			for(int i = 0; i < cantidad; i++) {
				cout << r[i] << endl;
			}
			cout << "¿Desea regresar al menú principal?" << "Si (1), Salir (2)" << endl;
			cin >> seguir;
			if (seguir == 2) {
				continua  = false;
			}else if(seguir == 1) {
				menu();
				cin >> opcion;
			}
			//cout << respuesta << endl;
		break;

		case 5:
			continua = false;
		break;

		default:
			cout << "opcion inválida" << endl;
			menu();
			cin >> opcion;
		break;
		}
	
	}
	return 0;
}

/**
* @brief Método que dado dos números obtiene el mínimo común multiplo.
* @param int num1, int num2: Números para obtener el mínimo común multiplo.
* @return int mcm: Mínimo común multiplo de dos valores.
*/
int mcm(int num1, int num2) {
    int mcm = 0;
    int a = std::max(num1, num2);
    int b = std::min(num1, num2);
    mcm = (a/mcd(a,b))*b;
    return mcm;
}

/**
* @brief Función auxiliar para correr las distintas funciones
* @param void
* @return void
*/
void menu() {
	cout << "Elige alguna opcion" << endl;
	cout << "[1] Calcular el mínimo común múltiplo" << endl;
	cout << "[2] Calcular el máximo común divisor" << endl;
	cout << "[3] Verificar si el último número de la sucesion de Collatz es uno " << endl;
	cout << "[4] lista de descomposición en factores primos de un número" << endl;
	cout << "[5] Salir" << endl;
}

/**
* @brief Método que dado dos números obtiene el maxímo común divisor.
* @param int num1, int num2: Valores numéricos.
* @return int mcd: Mínimo común multiplo de dos valores.
*/
int mcd(int num1, int num2) {
    int mcd = 0;
    int a = std::max(num1, num2);
    int b = std::min(num1, num2);
    while(b != 0) {
        mcd = b;
        b = a % b;
        a = mcd;
    }
    return mcd;
}

/**
* @brief Nos indica si el último número de la suceción de Collatz es uno.
* @param int n: Numero de la posición de la suceción de Collatz.
* @return bolean: Verdadero de es uno, falso si no lo es
*/
bool sucesionCollatz(int n) {
	while (n != 1) {
		if (n % 2 == 0)
			n = n / 2;
		else
			n = n * 3 +1;
		printf("%d ", n);
	}
	return (n == 1);
}

/**
* @brief Método que devuel la factorización de numeros primos de algún número.
* @param int n: Número a factorizar
* @return int[] lista: Factorización en números primos.
*/
int * factoresPrimos(int n) {
	int i = 2;
	static int r[100];
    cantidad = 0;
	while(n != 1) {
	    if (n % i == 0) {
		    n = n / i;
			r[cantidad ++] = i;
    	}else {
			i++;
		}  
  	}
	return r;
  	//return factores;
}